// storage-plugins/external-storage.js

const StorageLocator = function(externalStorege){
    const locations = externalStorege
    this.registerStorageLocation = (location) => locations[location] = {}
    this.storageLocation = (location) => {
        const loc = locations.get(location)
        if ( !loc ) throw new Error("Location not registered")
        return loc
    }
    this.getResourceName = ( location, resName, resType, options ) => {
        let res
        if ( ! locations[location][resName] ) locations[location][resName] = {}
        if ( ! locations[location][resName][resType] ) res = locations[location][resName][resType] = {}
        if ( options.resSubtype ){
            if ( ! locations[location][resName][resType][options.resSubtype] ) res = locations[location][resName][resType][options.resSubtype] = {}
        }

        return ( data ) => res.data = data
    }
}

const Storage = function(storageLocator) {

    this.registerStorageLocation = (location) => {
        storageLocator.registerStorageLocation(location)
    }

    this.storePkiResource = async function( location,resourceName,data ) {
        //console.log(`dump resource in the location [${location}] with key [${resourceName}] and value [${data.substring(0,25)} ... ${data.slice(-25)}]`)
        try {
            resourceName(data)
        } catch (error) {
            return Promise.reject(error)
        }
        return Promise.resolve()
    }

    this.dumpPkiResource = async function( location,resourceName ) {
        //console.log(`dump resource in the location [${location}] with key [${resourceName}]`)
        return Promise.resolve(storageLocator.storageLocation(location)[resourceName])
    }
}

export { Storage, StorageLocator }
